src/main/kotlin/io/emeraldpay/dshackle/rpc/NativeCall.kt

```diff
@@ -101,19 +104,60 @@ open class NativeCall(
     }
 
     open fun nativeCallResult(requestMono: Mono<BlockchainOuterClass.NativeCallRequest>): Flux<CallResult> {
+        val requestSpan = tracer.currentSpan()
         return requestMono.flatMapMany(this::prepareCall)
             .flatMap {
-                if (it.isValid()) {
-                    val parsed = parseParams(it.get())
-                    this.fetch(parsed)
-                        .doOnError { e -> log.warn("Error during native call: ${e.message}") }
+                val requestId = it.requestId
+                val requestCount = it.requestCount
+                val id = it.getContextId()
+                val result = if (it.isValid()) {
+                    run {
+                        val parsed = try {
+                            parseParams(it.get())
+                        } catch (e: Exception) {
+                            return@run Mono.error(e)
+                        }
+                        if (requestCount == 1 && requestId.isNotBlank()) {
+                            requestSpan?.tag("request.id", requestId)
+                        }
+                        this.fetch(parsed)
+                            .doOnError { e -> log.warn("Error during native call: ${e.message}") }
+                    }
                 } else {
                     val error = it.getError()
 
                     Mono.just(
                         CallResult(error.id, 0, null, error, null, null, null)
                     )
                 }
+
+                return@flatMap result
+                    .onErrorResume { err ->
+                        Mono.just(
+                            CallResult.fail(id, 0, err, null)
+                        )
+                    }
+                    .doOnNext { callRes ->
+                        if (requestCount > 1) {
+                            val span = tracer.currentSpan()
+                            if (callRes.isError()) {
+                                span?.error(
+                                    RuntimeException(callRes.error?.message ?: "Internal error")
+                                )
+                            }
+                            span?.end()
+                        }
+                    }
+                    .contextWrite { ctx ->
+                        if (requestCount > 1) {
+                            val span = tracer.nextSpan(requestSpan)
```

Comment:
The idea is there are 2 types of ids:
1. Id for the whole request 
2. Id for each request item.

Will we able to see the path of every item through the whole cycle?

Reply:
Yea, new span is created for every Call Item (if items size > 1) and because of that we can see path of every item. Now I use only Id for each request item and pass it to dshackle, and my idea was to get rid of mappings on dproxy side and pass also the whole request id in next iteration.